flink/src/test/java/org/apache/iceberg/flink/TestFlinkTableSource.java

```diff
@@ -90,17 +103,418 @@ public void testLimitPushDown() {
+    String sqlNotIn = String.format("SELECT * FROM %s WHERE id NOT IN (3,2) ", TABLE_NAME);
+    String explainNotIn = getTableEnv().explainSql(sqlNotIn);
+    String expectedFilter = "ref(name=\"id\") != 3,ref(name=\"id\") != 2";
+    Assert.assertTrue("explain should contains the push down filter", explainNotIn.contains(expectedFilter));
+
+    List<Object[]> resultNotIn = sql(sqlNotIn);
+    Assert.assertEquals("should have 1 record", 1, resultNotIn.size());
+    Assert.assertArrayEquals("Should produce the expected record", resultNotIn.get(0), new Object[] {1, "a"});
+    Assert.assertEquals("Should create only one scan", 1, scanEventCount);
+    Assert.assertEquals("should contains the push down filter", "(ref(name=\"id\") != 3 and ref(name=\"id\") != 2)",
+        lastScanEvent.filter().toString());
+
+    String sqlNotInNull = String.format("SELECT * FROM %s WHERE id NOT IN (1,2,NULL) ", TABLE_NAME);
+    String explainNotInNull = getTableEnv().explainSql(sqlNotInNull);
+    Assert.assertFalse("explain should not contains FilterPushDown",
+        explainNotInNull.contains(expectedFilterPushDownExplain));
+  }
+
+  @Test
+  public void testFilterPushDownIsNotNull() {
+    String sqlNotNull = String.format("SELECT * FROM %s WHERE data IS NOT NULL", TABLE_NAME);
+    String explainNotNull = getTableEnv().explainSql(sqlNotNull);
+    String expectedFilter = "not_null(ref(name=\"data\"))";
+    Assert.assertTrue("explain should contains the push down filter", explainNotNull.contains(expectedFilter));
+
+    List<Object[]> resultNotNull = sql(sqlNotNull);
+    Assert.assertEquals("should have 2 record", 2, resultNotNull.size());
+
+    List<Object[]> expected = Lists.newArrayList();
+    expected.add(new Object[] {1, "a"});
+    expected.add(new Object[] {2, "b"});
+    Assert.assertArrayEquals("Should produce the expected record", resultNotNull.toArray(), expected.toArray());
+
+    Assert.assertEquals("Should create only one scan", 1, scanEventCount);
+    Assert.assertEquals("should contains the push down filter", lastScanEvent.filter().toString(), expectedFilter);
+  }
+
+  @Test
+  public void testFilterPushDownIsNull() {
+    String sqlNull = String.format("SELECT * FROM %s WHERE data IS  NULL", TABLE_NAME);
+    String explainNull = getTableEnv().explainSql(sqlNull);
+    String expectedFilter = "is_null(ref(name=\"data\"))";
+    Assert.assertTrue("explain should contains the push down filter", explainNull.contains(expectedFilter));
+
+    List<Object[]> resultNull = sql(sqlNull);
+    Assert.assertEquals("should have 1 record", 1, resultNull.size());
+    Assert.assertArrayEquals("Should produce the expected record", resultNull.get(0), new Object[] {3, null});
+
+    Assert.assertEquals("Should create only one scan", 1, scanEventCount);
+    Assert.assertEquals("should contains the push down filter", lastScanEvent.filter().toString(), expectedFilter);
+  }
+
+  @Test
+  public void testFilterPushDownNot() {
```

Comment:
It looks like Flink is rewriting the expression to `notEq`. Is there a test case where the `not` is passed to Iceberg? Maybe using a complex expression like `not(or(eq("id", 1), eq("id", 2)))`

Reply:
I update it